source/opt/upgrade_memory_model.cpp

```diff
@@ -160,14 +160,36 @@ void UpgradeMemoryModel::UpgradeMemoryAndImages() {
       }
 
       switch (inst->opcode()) {
-        case spv::Op::OpLoad:
+        case spv::Op::OpLoad: {
+          Instruction* src_pointer = context()->get_def_use_mgr()->GetDef(
+              inst->GetSingleWordInOperand(0u));
+          analysis::Type* src_type =
+              context()->get_type_mgr()->GetType(src_pointer->type_id());
+          auto storage_class = src_type->AsPointer()->storage_class();
+          if (storage_class == spv::StorageClass::Function) {
+            // If the buffer from function variable, flag NonPrivatePointer is
+            // unnecessary.
+            is_coherent = false;
```

Comment:
Variable pointers can also be stored in the Private storage class.

Reply:
Thanks for pointing.
Now, fixed it, added testcases for private storage class, and squash these.